tests/test_reg_loss_integration.py

```diff
@@ -0,0 +1,108 @@
+
+TEST_CASES = [
+    [BendingEnergyLoss, {}, ["pred"]],
+    [
+        LocalNormalizedCrossCorrelationLoss,
+        {"in_channels": 3, "kernel_size": 5, "kernel_type": "rectangular"},
+        ["pred", "target"],
+    ],
+    [
+        LocalNormalizedCrossCorrelationLoss,
+        {"in_channels": 3, "kernel_size": 5, "kernel_type": "triangular"},
+        ["pred", "target"],
+    ],
+    [
+        LocalNormalizedCrossCorrelationLoss,
+        {"in_channels": 3, "kernel_size": 5, "kernel_type": "gaussian"},
+        ["pred", "target"],
+    ],
+    [GlobalMutualInformationLoss, {"num_bins": 5}, ["pred", "target"]],
+]
+
+
+class TestRegLossIntegration(unittest.TestCase):
+    def setUp(self):
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
+        torch.manual_seed(0)
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu:0")
+
+    def tearDown(self):
+        torch.backends.cudnn.deterministic = False
+        torch.backends.cudnn.benchmark = True
+
+    @parameterized.expand(TEST_CASES)
+    def test_convergence(self, loss_type, loss_args, forward_args):
+        """
+        The goal of this test is to assess if the gradient of the loss function
+        is correct by testing if we can train a one layer neural network
+        to segment one image.
+        We verify that the loss is decreasing in almost all SGD steps.
+        """
+        learning_rate = 0.001
+        max_iter = 40
+
+        # define a simple 3d example
+        target = torch.rand((1, 3, 5, 5, 5), device=self.device)
+        image = 12 * target + 27
+        image = image.to(device=self.device)
+
+        # define a one layer model
+        class OnelayerNet(nn.Module):
+            def __init__(self):
+                super(OnelayerNet, self).__init__()
+                self.layer = nn.Sequential(
+                    nn.Conv3d(in_channels=3, out_channels=3, kernel_size=3, padding=1),
+                    nn.ReLU(),
+                    nn.Conv3d(in_channels=3, out_channels=3, kernel_size=3, padding=1),
+                )
+
+            def forward(self, x):
+                return self.layer(x)
+
+        # initialise the network
+        net = OnelayerNet().to(self.device)
+
+        # initialize the loss
+        loss = loss_type(**loss_args).to(self.device)
+
+        # initialize a SGD optimizer
+        optimizer = optim.Adam(net.parameters(), lr=learning_rate)
+
+        # train the network
+        for _ in range(max_iter):
+            # set the gradient to zero
+            optimizer.zero_grad()
+
+            # forward pass
+            output = net(image)
+            loss_input = {"pred": output, "target": target}
+
+            loss_val = loss(**{k: loss_input[k] for k in forward_args})
+
+            # backward pass
+            loss_val.backward()
+            optimizer.step()
```

Comment:
shall we have some assertions here about the decreasing loss value? or they are omitted on purpose?

Reply:
there was a bug in the original implementation of `LocalNormalizedCrossCorrelation` which causes large fluctuations. The bug is fixed now and decreased loss assertion is added.